lib/minheap.js

```diff
@@ -38,26 +51,46 @@ class MinHeap {
   }
 
   // This method returns true if the heap is empty
-  // Time complexity: ?
-  // Space complexity: ?
+  // Time complexity: O(1)
+  // Space complexity: O(1)
   isEmpty() {
-    throw new Error("Method not implemented yet...");
+    return this.store.length == 0;
   }
 
   // This helper method takes an index and
   //  moves it up the heap, if it is less than it's parent node.
   //  It could be **very** helpful for the add method.
-  // Time complexity: ?
-  // Space complexity: ?
+  // Time complexity: O(log n)
+  // Space complexity: O(log n)
   heapUp(index) {
-    throw new Error("Method not implemented yet...");
+    let parent = Math.floor((index - 1)/2);
+    if (parent < 1) parent = 0;
+
+    if (this.store[parent].key > this.store[index].key) {
+      this.swap(parent, index);
+      return this.heapUp(parent);
+    }
   }
 
   // This helper method takes an index and 
   //  moves it up the heap if it's smaller
   //  than it's parent node.
   heapDown(index) {
-    throw new Error("Method not implemented yet...");
+    const leftChild = 2 * index + 1;
+    const rightChild = 2 * index + 2;
+
+    if (leftChild >= this.store.length || rightChild >= this.store.length) return;
```

Comment:
What if the left child is smaller than the length and the left child is smaller than it's parent, and at the same time the right child is equal to the length?

Reply:
@CheezItMan can you give me an example? I'm having a hard time thinking of how a child index would be smaller than its parent index.